### ISSUE TITLE
Add option to dump response content to a folder

### ISSUE THREAD

**Issue by alexander-schranz** [dantleech/fink#45]
I want to use fink not only to check for false response codes. I want also to dump its response content (in my case html) to a file because I want then to use the [w3c html5validator](https://github.com/validator/validator) to validate this files.

**Comment by dantleech** [comment]
_maybe_. My main concern would be that while dumping the HTML is easy enough, it leads on to dumping the HTML + assets.. and perhaps to making all URLs relative and creating an offline version of a given site.

I think Fink could also be used as a library, and you could f.e.do

```
DispatcherBuilder::create('http://www.example.com')->callback(function (Response $response) {

**Comment by dantleech** [comment]
Note that there is also https://github.com/spatie/crawler which might be more suitable for your use case?

Not against the idea necessarily, it would be convenient, but actually not 100% sure it belongs here (it could fit though)

**Comment by dantleech** [comment]
If we did, I guess the `Crawler` should be refactored to extract the DOM parsing into an observer, the code to dump the HTML can then also be an observer, and the Crawler will only send notifications when it gets a Response.

Not sure what the API would look like, but would probably want to do it after parsing the body.

**Comment by alexander-schranz** [comment]
> If we did, I guess the `Crawler` should be refactored to extract the DOM parsing into an observer, the code to dump the HTML can then also be an observer, and the Crawler will only send notifications when it gets a Response and has read the `$body`.

Throwing an Event sounds like a good idea for me and would make it very flexible: Would you use the symfony/event-dispatcher for this or which library do you prefer?

**Comment by dantleech** [comment]
No, no libraries for this :) We can simply create an interface for the observer (e.g. `CrawlerObserver`) and pass a collection of these to the crawler (e.g. `CrawlerObservers`).

**Comment by dantleech** [comment]
I think this will require some refactoring, still thinking about how to do it...
